### PR TITLE
FileContentProvider: use projections for helper queries

### LINKED ISSUE #5926: App crashes when opening folder that contains subfolder with many files
### Steps to reproduce

This is what I think happened. I haven't had the chance to try to reliably reproduce yet, but if someone wants me to, I could give it a try. Anyway, here goes:

1. Create a folder `foo` and a subfolder `bar` inside it
1. Create a huge number of files (4700+ in my case, doesn't have to be done on Android) in `bar`
1. Open `foo` and `bar` in the Android app so it's cached. 
1. On another device (or in the web GUI), create a new subfolder in `foo` called `baz` and move `bar` into `baz`. 
1. Open `foo` in the Android app again, and it crashes.

************ CAUSE OF ERROR ************

```
android.database.sqlite.SQLiteBlobTooBigException: Row too big to fit into CursorWindow requiredPos=3157, totalRows=1560
	at android.database.sqlite.SQLiteConnection.nativeExecuteForCursorWindow(Native Method)
	at android.database.sqlite.SQLiteConnection.executeForCursorWindow(SQLiteConnection.java:942)
	at android.database.sqlite.SQLiteSession.executeForCursorWindow(SQLiteSession.java:838)
	at android.database.sqlite.SQLiteQuery.fillWindow(SQLiteQuery.java:62)
	at android.database.sqlite.SQLiteCursor.fillWindow(SQLiteCursor.java:161)
	at android.database.sqlite.SQLiteCursor.onMove(SQLiteCursor.java:131)
	at android.database.AbstractCursor.moveToPosition(AbstractCursor.java:249)
	at android.database.AbstractCursor.moveToNext(AbstractCursor.java:281)
	at com.owncloud.android.providers.FileContentProvider.deleteDirectory(FileContentProvider.java:191)
	at com.owncloud.android.providers.FileContentProvider.delete(FileContentProvider.java:135)
	at com.owncloud.android.providers.FileContentProvider.deleteDirectory(FileContentProvider.java:185)
	at com.owncloud.android.providers.FileContentProvider.delete(FileContentProvider.java:135)
	at com.owncloud.android.providers.FileContentProvider.delete(FileContentProvider.java:115)
	at android.content.ContentProviderOperation.applyInternal(ContentProviderOperation.java:342)
	at android.content.ContentProviderOperation.apply(ContentProviderOperation.java:314)
	at com.owncloud.android.providers.FileContentProvider.applyBatch(FileContentProvider.java:671)
	at android.content.ContentProvider.applyBatch(ContentProvider.java:2179)
	at android.content.ContentProvider$Transport.applyBatch(ContentProvider.java:398)
	at android.content.ContentProviderClient.applyBatch(ContentProviderClient.java:532)
	at android.content.ContentProviderClient.applyBatch(ContentProviderClient.java:520)
	at android.content.ContentResolver.applyBatch(ContentResolver.java:1915)
	at com.owncloud.android.datamodel.FileDataStorageManager.saveFolder(FileDataStorageManager.java:407)
	at com.owncloud.android.operations.RefreshFolderOperation.synchronizeData(RefreshFolderOperation.java:478)
	at com.owncloud.android.operations.RefreshFolderOperation.fetchAndSyncRemoteFolder(RefreshFolderOperation.java:369)
	at com.owncloud.android.operations.RefreshFolderOperation.run(RefreshFolderOperation.java:228)
	at com.owncloud.android.lib.common.operations.RemoteOperation.run(RemoteOperation.java:357)
	at java.lang.Thread.run(Thread.java:919)

************ APP INFORMATION ************
ID: com.nextcloud.client
Version: 30110190
Build flavor: gplay

************ DEVICE INFORMATION ************
Brand: OnePlus
Device: OnePlus8Pro
Model: IN2023
Id: QKQ1.191222.002
Product: OnePlus8Pro_EEA

************ FIRMWARE ************
SDK: 29
Release: 10
Incremental: 2004111413
```

### Expected behaviour

The app shouldn't crash. 

### Actual behaviour

The app crashes over and over again when opening the `foo` folder. 

### Environment data
Android version: 10

Device model: OnePlus 8 Pro

Stock or customized system: stock

Nextcloud app version: 3.11.1

Nextcloud server version: 18.0.3

Reverse proxy: nginx

### PR DESCRIPTION
**Pull request by AlvaroBrey**
This limits the amount of data in the resulting cursor and helps prevent CursorTooBig exceptions.

Fixes #5926

- [x] Tests written, or not not needed

### CODE REVIEW

File: src/main/java/com/owncloud/android/providers/FileContentProvider.java [hunk 1]
```diff
@@ -100,6 +101,22 @@
     public static final int ARBITRARY_DATA_TABLE_INTRODUCTION_VERSION = 20;
     public static final int MINIMUM_PATH_SEGMENTS_SIZE = 1;
 
+    private static final String[] PROJECTION_CONTENT_TYPE = new String[]{ProviderTableMeta._ID, ProviderTableMeta.FILE_CONTENT_TYPE};
+    private static final String[] PROJECTION_REMOTE_ID = new String[]{ProviderTableMeta._ID, ProviderTableMeta.FILE_REMOTE_ID};
+    private static final String[] PROJECTION_FILE_AND_STORAGE_PATH = new String[]{ProviderTableMeta._ID, ProviderTableMeta.FILE_STORAGE_PATH, ProviderTableMeta.FILE_PATH};
```
Comment (tobiasKaminsky) [inline src/main/java/com/owncloud/android/providers/FileContentProvider.java:106]: line wrap

**Review by tobiasKaminsky — APPROVED** [review]

**Comment by tobiasKaminsky** [comment]
> @tobiasKaminsky Tested manually with the steps outlined in #5926, but would be good to test this in integration tests. Any pointers as to how to do that?

Some tests are in FileDataStorageManagerIT.

(More specifically what we need to do is: Create a directory containing 5000 files in the database, then call deleteDirectory on that directory)

That is the "problem" as this depends on Android/device, as allowed query size changed over time.
If we want to do write such a test it must only work on DB, not with server.

So something like
for (1.5000) {
storagemanager.saveFile(…)
}

**Comment by AlvaroBrey** [comment]
> So something like for (1.5000) { storagemanager.saveFile(…) }

I've been trying to implement something like this, but I can't reproduce the error in the same device where it crashes with the manual steps.

Additionally it takes way too long to execute, at least on a physical device.  So unless you  have other preference, we can merge this without further testing.

**Comment by AlvaroBrey** [comment]
> > So something like for (1.5000) { storagemanager.saveFile(…) }
> 
> I've been trying to implement something like this, but I can't reproduce the error in the same device where it crashes with the manual steps.
> 
> Additionally it takes way too long to execute, at least on a physical device. So unless you have other preference, we can merge this without further testing (as long as CI passes of course)

For reference, tested on Android 11 (LineageOS)

**Comment by Tanman897** [comment]
So @AlvaroBrey, am I reading this correctly? If we are having the problem on Nextcloud Android where folders with large amounts of photos are crashing the app, then we can download the above APK file and it will fix the issue? Why not just include it in a Nextcloud app update then?

**Comment by AlvaroBrey** [comment]
@Tanman897 It's going to be included in an upcoming update (in fact, a RC is coming this week if all goes according to plan). We can't just do a release for every merged PR or we'd spent all of our (already pretty limited) time releasing :\.

The QA APKs generated in PRs are just for testing purposes. We can point users to them and ask them to confirm if their issues are solved, for the cases where we can't reproduce the issue ourselves.

**Comment by Tanman897** [comment]
@AlvaroBrey Nice; I will be looking forward to the update. :)

**Pull request merged**